ofnetSvcProxy.go

```diff
@@ -0,0 +1,662 @@
+}
+
+// svcCatalogue holds information about all services to be proxied
+// Accessible by north-bound API
+type svcCatalogue struct {
+	svcMap map[string]ServiceSpec  // service name as key
+	provMap map[string]Providers   // service name as key
+}
+
+// provOper holds operational info for each provider
+type provOper struct {
+	clientEPs map[string]bool	// IP's of endpoints served by the provider
+	pqHdl *pqueue.Item	// handle into the providers pq
+}
+
+// proxyOper is operational state of the proxy
+type proxyOper struct {
+	ports []PortSpec
+	provHdl map[string]provOper  // provider IP as key
+	provPQ *pqueue.MinPQueue  // provider priority queue for load balancing
+	watchedFlows []*ofctrl.Flow  // flows this service is watching
+	natFlows map[string]*ofctrl.Flow // epIP.[in|out] as key
+}
+
+// ServiceProxy is an instance of a service proxy
+type ServiceProxy struct {
+	ofSwitch    *ofctrl.OFSwitch        // openflow switch we are talking to
+	dNATTable   *ofctrl.Table           // proxy dNAT rules table
+	sNATTable   *ofctrl.Table           // proxy sNAT rules table
+	dNATNext    *ofctrl.Table           // Next table to goto for dNAT'ed packets
+	sNATNext    *ofctrl.Table           // Next table to goto for sNAT'ed packets
+	catalogue   svcCatalogue		// Services and providers added to the proxy
+	oMutex sync.Mutex		// mutex between management and datapath
+	operState   map[string]*proxyOper	// Operational state info, with service IP as key
+}
+
+func getIPProto(prot string) uint8 {
+	switch prot {
+	case "TCP":
+		return ofctrl.IP_PROTO_TCP
+	case "UDP":
+		return ofctrl.IP_PROTO_UDP
+	default:
+		return 0
+	}
+}
+
+func matchSpec(s1, s2 *ServiceSpec) bool {
+	if s1.IpAddress != s2.IpAddress {
+		return false
+	}
+
+	if len(s1.Ports) != len(s2.Ports) {
+		return false
+	}
+
+	// if the order or ports changes, this will treat as a mismatch
+	// but, not a big deal...
+	for ix := 0; ix < len(s1.Ports); ix++ {
+		if s1.Ports[ix].Protocol != s2.Ports[ix].Protocol {
+			return false
+		}
+		if s1.Ports[ix].SvcPort != s2.Ports[ix].SvcPort {
+			return false
+		}
+		if s1.Ports[ix].ProvPort != s2.Ports[ix].ProvPort {
+			return false
+		}
+	}
+
+	return true
+}
+
+// allocateProvider gets the provider with least load
+// also updates the provider to client linkage
+func (svcOp *proxyOper)allocateProvider(clientIP string) (net.IP, error) {
+	if svcOp.provPQ.Len() <= 0 {
+		return net.ParseIP("0.0.0.0"), errors.New("No provider")
+	}
+	prov := svcOp.provPQ.GetMin()
+	svcOp.provPQ.IncreaseMin()
+	svcOp.provHdl[prov].clientEPs[clientIP] = true
+	return net.ParseIP(prov), nil
+}
+
+func getNATKey(epIP, natT string, p *PortSpec) string {
+	key := epIP + "." + natT + "." + p.Protocol + strconv.Itoa(int(p.SvcPort))
+	return key
+}
+
+// addNATFlow sets up a NAT flow
+// natT must be "Src" or "Dst"
+func (svcOp *proxyOper)addNATFlow(this, next *ofctrl.Table, p *PortSpec,
+				 ipSa, ipDa, ipNew *net.IP, natT string) {
```

Comment:
Because this is the data path, no meaningful action can be taken by a caller if the flow fails. The best we can do is log it (which the inner function already does).